internal/operator/cluster/cluster.go

```diff
@@ -413,6 +413,46 @@ func ScaleDownBase(clientset kubeapi.Interface, replica *crv1.Pgreplica, namespa
 
 }
 
+// UpdateAnnotations updates the annotations in the "template" portion of a
+// PostgreSQL deployment
+func UpdateAnnotations(clientset kubernetes.Interface, restConfig *rest.Config,
+	cluster *crv1.Pgcluster, annotations map[string]string) error {
+	var updateError error
+
+	// first, get a list of all of the instance deployments for the cluster
+	deployments, err := operator.GetInstanceDeployments(clientset, cluster)
+
+	if err != nil {
+		return err
+	}
+
+	// now update each deployment with the new annotations
+	for _, deployment := range deployments.Items {
+		log.Debugf("update annotations on [%s]", deployment.Name)
+		log.Debugf("new annotations: %v", annotations)
+
+		deployment.Spec.Template.ObjectMeta.SetAnnotations(annotations)
```

Comment:
Will this overwrite any existing annotations in the Pod?  At the moment specifically thinking about the `status` annotation written by Patroni, and not sure if any conflict we need to be concerned with:

```
$ kubectl get pods mycluster1-bc47bff86-k4rwg -o yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    status: '{"conn_url":"postgres://10.44.0.165:5432/postgres","api_url":"http://10.44.0.165:8009/patroni","state":"running","role":"master","version":"1.6.5","xlog_location":83886080,"timeline":1}'
```

Reply:
Does Patroni reference that status, or is that merely for convenience?

Reply:
I know that Patroni uses that annotation to obtain information about other instances in the cluster, e.g. when `patronictl list` is run.  To test this locally, I overwrote the status annotation in a pod and quickly ran `patronictl list`, which resulted in an error:

```
$ kubectl annotate pod mycluster1-uaql-7dccc96fc6-xcfxg status="something" --overwrite    
pod/mycluster1-uaql-7dccc96fc6-xcfxg annotated
[vagrant@client-centos crunchy-containers]$ kubectl exec -it mycluster1-bc47bff86-k4rwg -- patronictl list                      
Traceback (most recent call last):
  File "/usr/local/bin/patronictl", line 11, in <module>
    sys.exit(ctl())
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/patroni/ctl.py", line 829, in members
    output_members(cluster, cluster_name, extended, fmt)
  File "/usr/local/lib/python3.6/site-packages/patroni/ctl.py", line 763, in output_members
    len(set(m['host'] for m in cluster['members'])) < len(members)
  File "/usr/local/lib/python3.6/site-packages/patroni/ctl.py", line 763, in <genexpr>
    len(set(m['host'] for m in cluster['members'])) < len(members)
KeyError: 'host'
command terminated with exit code 1
```

However, right after that Patroni fixes the annotation and all starts working again.

```
[vagrant@client-centos crunchy-containers]$ kubectl exec -it mycluster1-bc47bff86-k4rwg -- patronictl list                      
+ Cluster: mycluster1 (6863113107911381134) -----+--------+---------+----+-----------+
|              Member              |     Host    |  Role  |  State  | TL | Lag in MB |
+----------------------------------+-------------+--------+---------+----+-----------+
|    mycluster1-bc47bff86-k4rwg    | 10.44.0.165 | Leader | running |  1 |           |
| mycluster1-uaql-7dccc96fc6-xcfxg | 10.44.0.168 |        | running |  1 |         0 |
+----------------------------------+-------------+--------+---------+----+-----------+
```

So my guess based on the above is that Patroni will just re-add it if we overwrite it when adding any custom annotations (and hopefully without Patroni clobbering any of our annotations when it is re-added).

Reply:
@andrewlecuyer So...the annotations are being added to the Deployment, which then get propagated to the Pod. I'm going to observe a bit of what happens to ensure everything is playing nicely.

Reply:
@andrewlecuyer OK, here is why it's OK. It is because, in fact, we are editing the annotations at the Deployment level, which get propagated to the Pod.

When the modifications take place, the Pod must be destroyed anyway, so regardless, Patroni will have to add a status label to it.

So this is fine.